app/src/main/kotlin/com/intive/picover/images/view/ImagesScreen.kt

```diff
@@ -18,27 +18,23 @@ import androidx.compose.ui.Modifier
 import androidx.compose.ui.graphics.Color
 import androidx.compose.ui.unit.dp
 import com.intive.picover.common.loader.PicoverLoader
-import com.intive.picover.images.viewmodel.ImagesStorageState
-import com.intive.picover.images.viewmodel.ImagesStorageState.Error
-import com.intive.picover.images.viewmodel.ImagesStorageState.Loading
-import com.intive.picover.images.viewmodel.ImagesStorageState.Success
+import com.intive.picover.common.viewmodel.state.ViewModelState
 import com.intive.picover.images.viewmodel.ImagesViewModel
 import com.skydoves.landscapist.coil.CoilImage
 
 @Composable
 fun ImagesScreen(viewModel: ImagesViewModel) {
-	val state by viewModel.uiState
+	val state by viewModel.state
 	ValidateFirebaseStorageState(state)
 }
 
 @Composable
-private fun ValidateFirebaseStorageState(state: ImagesStorageState) {
-	when (state) {
-		is Success -> {
-			val uris = state.uris
+private fun ValidateFirebaseStorageState(state: ViewModelState<List<Uri>>) {
+	when {
+		state.isLoaded() -> {
 			LazyColumn(
 				content = {
-					items(uris) {
+					items(state.data()!!) {
```

Comment:
Would be nice to have a smart cast here

Reply:
What a nice challenge :D 

Of course I can use `is Loaded` without function `isLoaded` to have no problems with casting  but i want to follow the DRY rule...

What i found out that works is:

```
@OptIn(ExperimentalContracts::class)
fun <T> ViewModelState<T>.isLoaded(): Boolean {
	contract {
		returns(true) implies (this@isLoaded is Loaded<T>)
	}
	return this is Loaded
}
```
But it looks like overkill for me :D

Reply:
Why you think it's an overkill?

Reply:
Mainly because i was not able to create a version of this function that is not an extension function of ViewModelState ( doing similar inside a class was a failure from my site) and it looks weird that we have two functions that are "normal" fuctions, for Loading and Error and this one, as a separate extension for Loaded.

But maybe it is not a big deal and it is better to use smart cast + get rid off data() function :D

Reply:
You have to elaborate more, what actually means "_doing similar inside a class was a failure from my site_"? Does the  failure means that it's not possible?

Reply:
yes, i wanted to create `isLoaded` as a class function, similar to `isLoading` and `isError` but could not create a function that works with smart casting without using `<T> ViewModelState<T>.` :/

Reply:
I've checked on my side and finally I see the point :) I'm resolving the issue to see what's others opinion here